### DISCUSSION TITLE
How do I auth user with username or email?

### DISCUSSION THREAD

**Discussion by nicitaacom** [supabase/supabase#16619] [Questions]
I have this form for login

In my form I have field 'email or username' but now I able to login with email - How do I enable login with email or username?
docs link or step by step guide will be preciate

![image](https://github.com/supabase/supabase/assets/39565703/802722c0-fbd9-461d-8482-277cc0cd207a)

### Here is a bit of code for my login function
```ts
async function login(e: React.FormEvent) {
    e.preventDefault()
    if (email && password) {
    try {
        const response = await supabase.auth.signInWithPassword({ email: email, password: password })
//other logic for login like set profile picture url
  }
}
```

**Comment by GaryAustin1** [comment]
There is no username login.  Username is not a standard field in Supabase auth.   I commented on your discord post with an idea using an rpc call to a postgres function with username and password to get the email and then login.  Not the greatest solution, but not sure if anyone has a better way.

**Reply by nicitaacom** [reply]
Anyway its increases UX that's why I relly preciate your answer and will do my best

**Comment by nicitaacom** [comment]
### I did it like so:
I use React + Vite + TypeScript + tailwind + zustand + supabase
```ts
async function login(e: React.FormEvent) {
    e.preventDefault()
    if (emailOrUsername && password) {
      const isEmail = emailOrUsername.includes("@")

      if (isEmail) {
        try {
          const response = await supabase.auth.signInWithPassword({ email: emailOrUsername, password: password })
          if (response.error) throw response.error

          if (response.data.user) {
            userStore.authUser(response.data.user.id)
            navigate(`/`, { replace: true })
            setSuccess("Success!")
          }
        } catch (error) {
          console.log(58, "error - ", error)
          setError("Error!")
        }
      } else {
        try {
          const { data, error } = await supabase.from("users").select("email").eq("username", emailOrUsername)
          if (error) throw error
          if (data) {
            const response = await supabase.auth.signInWithPassword({ email: data[0].email, password: password })
            if (response.error) throw response.error

            if (response.data.user) {
              userStore.authUser(response.data.user.id)
              setSuccess("You logged in!")
            } else {
              setError("user not found - made a typo?")
            }
          } else throw "LoginPage.tsx - no data"
        } catch (error) {
          console.error(78, "login - ", error)
          setError("Error")
        }
      }
    }
  }
  ```
  
  ### Explanations
  I have Input and state for this input
  If I have @ in my input with value={usmailOrUsername}
  I do manipulations for login as usual
  If its not I find in my table "users" column "email" that equals column "username" with row emailOrUsername (here is I know that in input no @ that's why its username)
  
  Of course when I register new user I insert data in 3 columns
  ```tsx
  const { error } = await supabase
            .from("users")
            .insert({ id: response.data.user.id, username: usernameValue, email: email })
          if (error) throw error
  ```
  
  <br/>
  
  Thank you @GaryAustin1 for idea how do I realize it

**Reply by Patrity** [reply]
hey @nicitaacom - I know it's been quite some time since this post - I was looking for a similar solution, however, it seems like your solution would expose a mapping of all usernames/emails to the public, no? If so, I would think this is a pretty glaring unintentional security flaw.

**Reply by nicitaacom** [reply]
> expose a mapping of all usernames/emails to the public, no?

This code
```ts
 const { data, error } = await supabase.from("users").select("email").eq("username", emailOrUsername)
 ```
 
 Could be runned in API route or server action (in case Next.js)
 Or in server.ts (in case React + Vite + node.js + express)
 
 supabase its libs/supabaseAdmin.ts
 ```ts
 import { Database } from "@/interfaces/types_db"
import { createClient } from "@supabase/supabase-js"

export const supabaseAdmin = createClient<Database>(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY,
)

export default supabaseAdmin
```

**Reply by Patrity** [reply]
I agree with that, but then the email is still passed to the frontend to authenticate, right? So if someone was targeting a specific username, they could get whatever email they wanted.

**Reply by nicitaacom** [reply]
> I agree with that, but then the email is still passed to the frontend to authenticate, right? 

Of course you enter email on frontend part (in input field)

**Reply by Patrity** [reply]
Right.... so if I wanted to find out what your email was, I could just enter `nicitaacom` and a random password I know will fail, then wait for your server-side response to come back and I have your email.
Email addresses should be protected unless users explicitly want them shared.

**Reply by nicitaacom** [reply]
Don't get response from server about email (otherwise send evidance how do you get my email)
You just get response `AuthError` or `User`

Checking is username match some email in DB happening in server action or API route - so you don't have access to data on a server side

**Reply by Patrity** [reply]
Right but the authentication call to `supabase.auth.signInWithPassword` has to take place on the client.
This diagram maybe explains the flow better. If you can provide a link to where you have this deployed, along with a valid username, or better yet a code sample, I could show you.
Otherwise if you are sure that this is not a possibility, I would love to see how, because I would really love username sign ins

![SS2024-05-16-18 30 09](https://github.com/supabase/supabase/assets/42813882/28cd1f06-b7c6-42b8-90af-155b71c8c893)

**Reply by nicitaacom** [reply]
1. On client you provide email or username (check if .includes("@") - then it email - otherwise its username - based on that you userstand what you need to do)
2. Then you may use redis (to rate limit requests from some IP) - I think Supabase have build in restrictions of ammount of tries to login
3. Then you watch video - I explained that it happens on server - so you may throw message like "Incorrect email or password" or if you want "User with this username doesn't exist" (its on your choise but I prefer second variant)

https://github.com/supabase/supabase/assets/39565703/4ed31fa2-d72d-4272-b2bc-7ec81a42e52a

**Reply by therealsujitk** [reply]
I've been looking for a solution to this as well, and @Patrity is right; exposing user emails isn't the right approach.

1. **Client-side Email Exposure:** If the client sends a request to Supabase with the username and gets the email in response, this would expose emails in the browser. Anyone could use a username to retrieve the associated email or even script multiple requests to harvest emails.
2. **Server-side Request with `ANON_KEY`:** @nicitaacom, I understand you're suggesting making the email fetching request server-side with Next.js or a similar library supporting SSR. While this is slightly safer since the email doesn't reach the browser, it could still be vulnerable if the `ANON_KEY` is used. An attacker could potentially extract the `ANON_KEY` from your site and use it to make email fetch requests from their own Supabase client.

<hr>

### Alternative Approaches

1. **Server-side Request with `SERVICE_KEY`:** Create a `public.users` table containing only usernames, with a foreign key reference to the `auth.users` table. This setup ensures that emails cannot be fetched using the `ANON_KEY`. You will then need to Implement a server-side endpoint that takes in a username and password, fetches the email using the `SERVICE_KEY`, uses that email to authenticate the user, all on the server side.
2. **Deno Edge Function:** Use a Deno edge function to perform the same task as the server-side approach. This method is particularly useful if you cannot execute server-side code directly.

> [!Important]
> Remember, the `SERVICE_KEY` should never be exposed, so the requests involving it MUST be made on the server side.

In both methods, you need to return the session to the client. On the client side, you'll then need to set the Supabase session using the response.

This is what I could think of, and all this is just theoretical, I haven't tried implementing it yet. I'll be sure to add some code here if I do implement it or think of a better solution. If anyone else has other ideas, I'd love to hear them too.

**Comment by therealsujitk** [comment]
I wrote a detailed guide on how to do this securely --> https://stackoverflow.com/questions/78550922/how-do-i-authorise-users-with-username-in-supabase

**Reply by nicitaacom** [reply]
`const supabaseClient`

I would name it `const supabaseAdmin` becuase it has service role key

So you will understand that `supabaseClient` its SDK with anon key and `supabaseAdmin` its supabase SDK with service role key

**Reply by nicitaacom** [reply]
Also auth will now work if you don't use headers from next.js

I mean if you will use supabaseClient or supabaseAdmin it will not work because you need to set cookie

I think with code will be more understandable

**This will work I mean set cookies**
```ts
import { createServerComponentClient } from "@supabase/auth-helpers-nextjs"
import { cookies } from "next/headers"
import { Database } from "@/interfaces/types_db"

const supabaseServer = () => {
  cookies().getAll() // Keep cookies in the JS execution context for Next.js build
  return createServerComponentClient<Database>({ cookies })
}

export default supabaseServer

```

**This will not work** (because it don't set cookies)
```ts
import { Database } from "@/interfaces/types_db"
import { createClient } from "@supabase/supabase-js"

export const supabaseAdmin = createClient<Database>(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.SUPABASE_SERVICE_ROLE_KEY,
)

export default supabaseAdmin
```

**Reply by therealsujitk** [reply]
@nicitaacom oh yeah you're right, I didn't need to add the header. Also, the code I've written is a Deno edge function, so it isn't dependent on any SSR framework, It can by used anywhere.

**Reply by SheepTD** [reply]
Hi Sujitk, I love your edge function solution and I am planning on implementing it in my current project. I've seen that you included a curl request to test the edge function, but I was hoping you could explain how to send a request using JavaScript so it can be used via a button in the front end? Thanks.

**Reply by therealsujitk** [reply]
@SheepTD you can call in edge function using SupabaseJS as described [here](https://supabase.com/docs/reference/javascript/functions-invoke). So it would look like this,

```js
const { data, error } = await supabase.functions.invoke('login', {
  body: {
    identifier: 'username',
    password: 'password'
  }
});
```

If you aren't using SupabaseJS, you can call it like you would [any other POST request](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch).

> [!NOTE]
> If you're calling this function from your browser, you'll need to add CORS headers. You can [refer these docs](https://supabase.com/docs/guides/functions/cors) to learn how.

**Comment by Arzzam** [comment]
In this case we might to have the email id for the specific user in our db right? 
Is that possible to not have a email, In my case I just want to have a username and password login without the email? 
or having a dummy email for the specific user (let's assume username1@supabase.com) by turning off the email verification, even though I did this, I was wondering what if I forgot the password?

Any suggestion how I can implement this cases?
